### PR TITLE
[Branch 4.6] compaction_manager: Fix race when selecting sstables for rewrite operations

### LINKED ISSUE #8168: Potential data resurrection due to lack of synchronization in cleanup compaction
Data resurrection can happen if table T on node A still has non-owned data after cleanup finished, and that same node A regains ownership of that data in table T.

Node A, in the example above, can be left with non-owned data because:
1) memtable is not flushed prior to cleanup, so whatever non-owned data is in a memtable will eventually be flushed into an sstable (#1239).
2) cleanup doesn't wait for regular compaction, so it may miss any candidate containing non-owned data being currently compacted.

To fix them, we need to, respectively:
1) make sure memtable is flushed for a given table before starting cleanup on it (#1239)
2) wait for regular compaction(s) completion before finding candidates for cleanup, then these candidates will be marked as compacting by compaction manager and cannot be later picked as candidates by regular compaction.

### PR DESCRIPTION
**Pull request by raphaelsc**
Rewrite operations are scrub, cleanup and upgrade.

Race can happen because 'selection of sstables' and 'mark sstables as
compacting' are decoupled. So any deferring point in between can lead
to a parallel compaction picking the same files. After commit 2cf0c4bbf,
files are marked as compacting before rewrite starts, but it didn't
take into account the commit c84217ad which moved retrieval of
candidates to a deferring thread, before rewrite_sstables() is even
called.

Scrub isn't affected by this because it uses a coarse grained approach
where whole operation is run with compaction disabled, which isn't good
because regular compaction cannot run until its completion.

From now on, selection of files and marking them as compacting will
be serialized by running them with compaction disabled.

Now cleanup will also retrieve sstables with compaction disabled,
meaning it will no longer leave uncleaned files behind, which is
important to avoid data resurrection if node regains ownership of
data in uncleaned files.

Fixes #8168.
Refs #8155.

[backport notes:
- minor conflict around run_with_compaction_disabled()
- bumped into our old friend
  https://gcc.gnu.org/bugzilla/show_bug.cgi?id=95111,
so I had to use std::ref() on local copy of lambda.
]

Signed-off-by: Raphael S. Carvalho <raphaelsc@scylladb.com>
Message-Id: <20211129133107.53011-1-raphaelsc@scylladb.com>
(cherry picked from commit 80a1ebf0f37642d3a64c8098308339d16f43730a)

Signed-off-by: Raphael S. Carvalho <raphaelsc@scylladb.com>

### CODE REVIEW

**Comment by denesb** [comment]
> Fixes #8168.

Why is this fixing a closed issue?

File: compaction/compaction_manager.cc [hunk 1]
```diff
@@ -873,31 +884,29 @@ future<> compaction_manager::perform_cleanup(database& db, column_family* cf) {
         return make_exception_future<>(std::runtime_error(format("cleanup request failed: there is an ongoing cleanup on {}.{}",
             cf->schema()->ks_name(), cf->schema()->cf_name())));
     }
-    return seastar::async([this, cf, &db] {
+  // FIXME: indentation
+  auto sorted_owned_ranges = db.get_keyspace_local_ranges(cf->schema()->ks_name());
+  auto get_sstables = [this, &db, cf, sorted_owned_ranges] () -> future<std::vector<sstables::shared_sstable>> {
+    return seastar::async([this, &db, cf, sorted_owned_ranges = std::move(sorted_owned_ranges)] {
         auto schema = cf->schema();
-        auto sorted_owned_ranges = db.get_keyspace_local_ranges(schema->ks_name());
```
Comment (denesb) [inline compaction/compaction_manager.cc:878]: Why was this moved out of the thread?

File: compaction/compaction_manager.hh [hunk 2]
```diff
@@ -178,7 +178,7 @@ private:
     maintenance_scheduling_group _maintenance_sg;
     size_t _available_memory;
 
-    using get_candidates_func = std::function<std::vector<sstables::shared_sstable>(const column_family&)>;
+    using get_candidates_func = std::function<future<std::vector<sstables::shared_sstable>>()>;
```
Comment (denesb) [inline compaction/compaction_manager.hh:181]: Is a future really necessary? I see one getter using a thread by AFAU it is not needed either.

File: compaction/compaction_manager.cc (hunk 1 above)
Comment (raphaelsc) [reply compaction/compaction_manager.cc:878]: That's a backport to 4.6. It was moved out because we need to pass it to compaction_type_options, and also reference it inside the lambda for filtering out sstables that don't need cleanup

File: compaction/compaction_manager.hh (hunk 2 above)
Comment (raphaelsc) [reply compaction/compaction_manager.hh:181]: > Is a future really necessary? I see one getter using a thread by AFAU it is not needed either.

The reason it is needed is that cleanup impl of this func can potentially stall, so we have yielding points there.

**Comment by raphaelsc** [comment]
> > > Fixes #8168.
> > 
> > 
> > Why is this fixing a closed issue?
> 
> It's a backport.
> 
> Is 5.0 not vulnerable?

5.0 is not vulnerable. A test case failed with the backport patch. Once I push force the fix, I'll let you know.

**Comment by raphaelsc** [comment]
ping

**Comment by raphaelsc** [comment]
> ping

ping merge.

**Comment by raphaelsc** [comment]
Was merged in July 2022. Closing...
